### DISCUSSION TITLE
vertex description for color sent as VK_FORMAT_R8G8B8A8_UNORM but consumed as vec4

### DISCUSSION THREAD

**Discussion by ilopit** [ocornut/imgui#6049] [New Users Build/Link/Run issues ONLY!]
Hey, wise people of ImGui stupid question. 

In vulkan backend (imgui_impl_vulkan.cpp:833) you send color as VK_FORMAT_R8G8B8A8_UNORM (32bit int) but consume it as vec4 (4 x 32 bit float). Is it correct? I can't find info that glsl auto cast VK_FORMAT_R8G8B8A8_UNORM  to vec4 .

**Comment by PathogenDavid** [comment]
Not a stupid question! This is something that gets glossed over in graphics tutorials so you're not wrong to question it. (Although it's definitely a Vulkan question, not a Dear ImGui one 😅)

> VK_FORMAT_R8G8B8A8_UNORM (32bit int)

The `_UNORM` is important here. It's short for "unsigned normalized integer".

In this context each channel of the color is an 8-bit unsigned normalized integer. (There are other sizes too, like `VK_FORMAT_R16G16B16A16_UNORM` which is 16 bits per channel.)

The basic idea is that a bit pattern of all 0's represents the floating point value of `0.0f`. A pattern of all 1's (so 255 in this case) represents a value of `1.0f`.

Intermediate binary values represent intermediate floating point values, so `127` in this case is around `0.5f`.

This is defined in the Vulkan specification in [§33.1.2 Identification of Formats](https://registry.khronos.org/vulkan/specs/1.0/html/chap33.html#_identification_of_formats) (although IMO the [Direct3D definition](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#3.2.3.1%20Terminology) is easier to understand.)

Don't worry if the spec feels overwhelming, BTW. A lot of people never even reference it. It contains a ton of low level details you really don't need to worry about unless you work for one of the GPU manufacturers. I'm linking it so you don't have to take my word for it.

The takeaway here is that the channels of `UNORM` colors aren't integer values or even floating point values. They're their own thing, but on the GPU side we generally interact with them as floating point values. (From [a shader perspective](https://registry.khronos.org/vulkan/specs/1.0/html/chap33.html#_identification_of_formats), they're considered floats.)

> I can't find info that glsl auto cast VK_FORMAT_R8G8B8A8_UNORM to vec4 .

Vertex attributes and texture samples are both automatically converted from their source format to what the shader asks for.

For vertex attributes, this is defined in [§21.3.1 Vertex Input Extraction](https://registry.khronos.org/vulkan/specs/1.0/html/chap21.html#fxvertex-input-extraction). It's a little obtuse, but here's the important bit:

> For attributes that are not 64-bit data types, each component is converted to the format of the input variable based on its type and size[...]

For textures this is defined in [§16.3.2 Format Converison](https://registry.khronos.org/vulkan/specs/1.0/html/chap16.html#textures-format-conversion).
